pkg/flow/internal/controller/loader.go

```diff
@@ -483,8 +476,12 @@ func (l *Loader) evaluate(logger log.Logger, bn BlockNode) error {
 		l.cache.CacheArguments(c.ID(), c.Arguments())
 		l.cache.CacheExports(c.ID(), c.Exports())
 	case *ArgumentConfigNode:
-		if _, ok := l.cache.moduleArguments[c.Label()]; !ok && c.Optional() {
-			l.cache.CacheModuleArgument(c.Label(), c.Default())
+		if _, found := l.cache.moduleArguments[c.Label()]; !found {
+			if c.Optional() {
+				l.cache.CacheModuleArgument(c.Label(), c.Default())
+			} else {
+				err = fmt.Errorf("missing required argument %q to module", c.Label())
+			}
```

Comment:
Thanks for prototyping the change with removing the ConfigNodeMap. I definitely missed that we needed to do a search rather than just a get, so I think the ConfigNodeMap is good as it was. 

That said, I do prefer this approach here for checking for required arguments; can we revert this commit but keep this change in before merging?

Reply:
I reverted and reapplied the parts we wanted to keep.